File: 3-third-year-2023-2024/Internet-services-architecture-and-development/Exams/soluzione_compito_20230110_v20231218/client-td-connreuse.c
#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <netdb.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "rxb.h"
#include "utils.h"
#ifdef USE_LIBUNISTRING
#include <unistring.h>
#endif /* USE_LIBUNISTRING */

#define MAX_RESPONSE_SIZE (16 * 1024)


/* leggi_dati_sensori hostname porta */
/* argv[0]            argv[1]  argv[2] */
int main(int argc, char *argv[])
{
	rxb_t rxb;
	int err, sd;
	struct addrinfo hints, *res, *ptr;

	/* Controllo errori */
	if (argc != 3) {
		fprintf(stderr, "Error! Usage:\n\t%s hostname porta\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	/* Ripristiniamo semantica sensata pipe e socket */
	signal(SIGPIPE, SIG_IGN);

	/* Preparo chiamata getaddrinfo */
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;

	char *hostname = argv[1];
	char *service = argv[2];

	err = getaddrinfo(hostname, service, &hints, &res);
	if (err != 0) {
		fprintf(stderr, "Errore getaddrinfo: %s\n", gai_strerror(err));
		exit(EXIT_FAILURE);
	}

	/* Procedura di connessione con fallback */
	for (ptr = res; ptr != NULL; ptr = ptr->ai_next) {
		sd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
		if (sd < 0)
			continue;

		err = connect(sd, ptr->ai_addr, ptr->ai_addrlen);
		if (err == 0) 
			break;

		close(sd);
	}

	if (ptr == NULL) {
		fprintf(stderr, "Errore di connessione!\n");
		exit(EXIT_FAILURE);
	}
	
	/* Libero memoria allocata da getaddrinfo */
	freeaddrinfo(res);

	/* Alloco buffer rxb */
	rxb_init(&rxb, MAX_RESPONSE_SIZE * 2);

	/* Ciclo di richieste di servizio */
	for(;;) {
		char tipo_sensore[4096];
		char stanza_sensore[4096];
		char numero_letture[4096];

		puts("Inserisci tipo sensore (fine per terminare):");
		if (fgets(tipo_sensore, sizeof(tipo_sensore), stdin) == NULL) {
			fprintf(stderr, "Errore fgets tipo_sensore\n");
			exit(EXIT_FAILURE);
		}

		if (strcmp(tipo_sensore, "fine\n") == 0) {
			break;
		}

		puts("Inserisci stanza sensore (fine per terminare):");
		if (fgets(stanza_sensore, sizeof(stanza_sensore), stdin) == NULL) {
			fprintf(stderr, "Errore fgets stanza_sensore\n");
			exit(EXIT_FAILURE);
		}

		if (strcmp(stanza_sensore, "fine\n") == 0) {
			break;
		}

		puts("Inserisci numero letture (fine per terminare):");
		if (fgets(numero_letture, sizeof(numero_letture), stdin) == NULL) {
			fprintf(stderr, "Errore fgets numero_letture\n");
			exit(EXIT_FAILURE);
		}

		if (strcmp(numero_letture, "fine\n") == 0) {
			break;
		}

		err = write_all(sd, tipo_sensore, strlen(tipo_sensore));
		if (err < 0) {
			perror("write_all tipo_sensore");
			exit(EXIT_FAILURE);
		}
		
		err = write_all(sd, stanza_sensore, strlen(stanza_sensore));
		if (err < 0) {
			perror("write_all stanza_sensore");
			exit(EXIT_FAILURE);
		}

		err = write_all(sd, numero_letture, strlen(numero_letture));
		if (err < 0) {
			perror("write_all numero_letture");
			exit(EXIT_FAILURE);
		}

		/* Leggo risposta del server */
		while(1) {
			char line[4096];
			size_t line_len;

			/* Leggo riga di testo dal Server */
			memset(line, 0, sizeof(line));
			line_len = sizeof(line) - 1;
			err = rxb_readline(&rxb, sd, line, &line_len);
			if (err < 0) {
				perror("Errore rxb_readline");
				exit(EXIT_FAILURE);
			}

#ifdef USE_LIBUNISTRING
			if (u8_check((uint8_t *)line, line_len) != NULL) {
				fputs("Errore: testo UTF-8 non valido!", stderr);
				exit(EXIT_FAILURE);
			}
#endif /* USE_LIBUNISTRING */

			/* Esco in caso di riga di fine risposta */
			if (strcmp(line, "--- END RESPONSE ---") == 0)
				break;

			/* Stampo riga a video */
			puts(line);
		}
	}

	/* Libero buffer rxb */
	rxb_destroy(&rxb);

	/* Chiudo la socket */
	close(sd);

	return 0;
}

File: Internet-services-architecture-and-development/Exams/compito_20230110/soluzione_compito_20230110_v20231218/ClientTDConnreuse.java
import java.io.*;
import java.net.*;

public class ClientTDConnreuse {
    // java ClientTDConnreuse hostname port
    //                        args[0]  args[1]
    public static void main(String[] args) {
        // Controllo argomenti
        if (args.length != 2) {
            System.err.println("Error! Usage is\n\tjava ClientTDConnreuse hostname port");
            System.exit(1);
        }

        try {
            // Creo socket
            var s = new Socket(args[0], Integer.parseInt(args[1]));
            var netIn = new BufferedReader(new InputStreamReader(s.getInputStream(), "UTF-8"));
            var netOut = new BufferedWriter(new OutputStreamWriter(s.getOutputStream(), "UTF-8"));
            // Non devo specificare codifica UTF-8: uso codifica di default della piattaforma
            var keyboard = new BufferedReader(new InputStreamReader(System.in)); 

            for (;;) {
                System.out.println("Inserisci tipo sensore (fine per terminare):");
                var tipoSensore = keyboard.readLine();
                if (tipoSensore.equals("fine")) {
                    System.out.println("Ok, esco!");
                    break;
                }

                System.out.println("Inserisci stanza sensore (fine per terminare):");
                var stanzaSensore = keyboard.readLine();
                if (stanzaSensore.equals("fine")) {
                    System.out.println("Ok, esco!");
                    break;
                }

                System.out.println("Inserisci numero letture (fine per terminare):");
                var numeroLetture = keyboard.readLine();
                if (numeroLetture.equals("fine")) {
                    System.out.println("Ok, esco!");
                    break;
                }

                // Mando dati al Server
                netOut.write(tipoSensore); netOut.newLine();
                netOut.write(stanzaSensore); netOut.newLine();
                netOut.write(numeroLetture); netOut.newLine();
                netOut.flush();

                // Leggo risposta
                String line;
                while ((line = netIn.readLine()) != null) {
                    if (line.equals("--- END REQUEST ---"))
                        break;

                    System.out.println(line);
                }
            }

            s.close();
        }
        catch(Exception e) {
            System.err.println(e.getMessage());
            e.printStackTrace(System.err);
            System.exit(1);
        }
    }
}
